test/parallel/test-promises-unhandled-rejections.js

```diff
@@ -0,0 +1,591 @@
+    assert.strictEqual(p, promise);
+  });
+  var p = Promise.reject(e);
+  setImmediate(function() {
+    Promise.resolve().then(function () {
+      p.catch(function(){});
+    });
+  });
+});
+
+asyncTest(function(done) {
+  onUnhandledSucceed(done, function(reason, promise) {
+    assert.strictEqual(undefined, reason);
+    assert.strictEqual(p, promise);
+  });
+  setImmediate(function() {
+    Promise.resolve().then(function () {
+      Promise.resolve().then(function () {
+        Promise.resolve().then(function () {
+          Promise.resolve().then(function () {
+            p.catch(function(){});
+          });
+        });
+      });
+    });
+  });
+  var p = Promise.reject();
+});
+
+asyncTest(function(done) {
+  onUnhandledSucceed(done, function(reason, promise) {
+    assert.strictEqual(undefined, reason);
+    assert.strictEqual(p, promise);
+  });
+  var p = Promise.reject();
+  setImmediate(function() {
+    Promise.resolve().then(function () {
+      Promise.resolve().then(function () {
+        Promise.resolve().then(function () {
+          Promise.resolve().then(function () {
+            p.catch(function(){});
+          });
+        });
+      });
+    });
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledFail(done);
+  var a = Promise.reject(e);
+  process.nextTick(function() {
+    Promise.resolve().then(function() {
+      process.nextTick(function() {
+        Promise.resolve().then(function() {
+          a.catch(function() {});
+        })
+      })
+    })
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledFail(done);
+  setImmediate(function() {
+    var a = Promise.reject(e);
+    process.nextTick(function() {
+      Promise.resolve().then(function() {
+        process.nextTick(function() {
+          Promise.resolve().then(function() {
+            a.catch(function() {});
+          })
+        })
+      })
+    });
+  });
+});
+
+asyncTest(function(done) {
+  var d = domain.create();
+  var domainReceivedError;
+  d.on('error', function(e) {
+    domainReceivedError = e;
+  });
+  d.run(function() {
+    var e = new Error('error');
+    var domainError = new Error('domain error');
+    onUnhandledSucceed(done, function(reason, promise) {
+      assert.strictEqual(reason, e);
+      assert.strictEqual(domainReceivedError, domainError);
+      d.dispose();
+    });
+    var a = Promise.reject(e);
+    process.nextTick(function() {
+      throw domainError;
+    });
+  });
+});
+
+asyncTest(function(done) {
+  clean();
+  var e = new Error('error');
+  var expectedDiff;
+  var now = Date.now();
+  process.on('unhandledRejection', function(reason, promise) {
+    var now = Date.now();
+    var order = [];
+    process.nextTick(function() {
+      order.push(1);
+    });
+    setTimeout(function() {
+      order.push(2);
+      assert.deepEqual([1,2], order);
+      done();
+    }, 1);
+  });
+  Promise.reject(e);
+});
+
+asyncTest(function(done) {
```

Comment:
Throwing an error inside a rejectionHandled handler goes to unhandledException, and does not cause .catch() to throw an exception